### DISCUSSION TITLE
@micropython.native or mpy

### DISCUSSION THREAD

**Discussion by beetlegigg** [micropython/micropython#9225] [Using MicroPython] [answered]
For a class I have I can decorate it with @micropython.native or I can cross-compile it to a mpy file (without the decorator).   I think I could also create an mpy file that also has the  decorator but then I have to put --march=armv6m (which I think is correct for the rpi pico) but then I get 
Implementation specific options:
  emit={bytecode,native,viper} -- set the default code emitter
  heapsize=<n> -- set the heap size for the GC (default 2097152)
to which I supposed the answer to emit is --emit=native, but not too sure what to put as heapsize.  I did try --heapsize=2097152 but I just keep getting the Implementation specific message as above and no mpy file is created.

But nevertheless, leaving to one side for the moment the mpy with decorator scenario,  I can easily decorate my class or create an mpy.  Both appear to work OK.   Is there any guidance as to which might be the better choice?

**Comment by robert-hh** [comment, marked as answer]
If you ask about runtime speed, there is not difference between a .py and a .mpy version. It is just a when and where the converting from .py to .mpy happens. Standard code and @micropython.native  code have a difference in code execution speed. As a rule of thumb, @micropython.native code runs twice as fast.
So it is not a question of .mpy or @micropython.native code

**Comment by jimmo** [comment]
You don't need to specify the heap size when using mpy-cross. Only the architecture (and even then only if your code uses `@native`).

If you use emit=native with mpy-cross then it's like implicitly putting `@native` on all methods. This is almost never practically useful because you should only selectively use the native emitter for performance critical functions as the generated native code takes up more RAM than the equivalent bytecode.

As Robert said there's no runtime speed difference between bytecode generated on the device vs via mpy-cross in a .mpy file. (They use exactly the same code... mpy-cross is actually just micropython running on your PC in a special mode that can serialise the generated bytecode).

The reason to use a .mpy file is that:
- it will load faster (skips the compilation step)
- uses less (temporary) memory during loading (but has the same final resident size)
- uses less space in the filesystem (typically less than half the size of the original .py file, even less depending on which optimisation level you set).
